### PR TITLE
Issue #7763: Update AbstractChecks to log DetailAST - TrailingComment

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #7763: Update AbstractChecks to log DetailAST - TrailingComment

from Roman: it is not implementation that I would like to have in this Check ideally, but let move refactoring to new issue.
https://github.com/checkstyle/checkstyle/issues/9009
https://github.com/checkstyle/checkstyle/issues/9010

Diff Regression projects: https://gist.githubusercontent.com/HuGanghui/709e0a8266eb1656488051b4cfc9161f/raw/5731ab3b42c995e5823e02a34a00c25241bcec1d/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/HuGanghui/d3b892a11fa52287ec1723246f12718f/raw/9576b09b677ea0cf838aa3ae451519597efd46ca/TrailingComment.xml

### CODE REVIEW

**Comment by HuGanghui** [comment]
Github, generate report

**Comment by HuGanghui** [comment]
Github, generate report

**Comment by romani** [comment]
@HuGanghui, please take a look at other PR for example of report generation, you my to provide links to configs
Instructions https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation

**Comment by HuGanghui** [comment]
CI error is that I introduce new check, but I am not:

Diff is detected.
283PR Description grepped:
284You introduce new Check
285--- .ci-temp/verify-no-exception-configs/web.txt	2020-08-23 17:15:34.837125664 +0000
286+++ .ci-temp/verify-no-exception-configs/file.txt	2020-08-23 17:15:34.845125665 +0000
287@@ -139,7 +139,6 @@
288 RegexpOnFilename
289 RegexpSingleline
290 RegexpSinglelineJava
291-RequireEmptyLineBeforeBlockTagGroup
292 RequireThis
293 ReturnCount
294 RightCurly
295Please create PR to repository https://github.com/checkstyle/contribution
296and add your new Check 
297   to file checkstyle-tester/checks-nonjavadoc-error.xml
298or to file checkstyle-tester/checks-only-javadoc-error.xml
299PR for contribution repository will be merged right after this PR.

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
@timurt ,

do you see why I can not match comment by content ? 

 ```
$ cat Test.java 
package org.checkstyle.suppressionxpathfilter.trailingcomment;

public class SuppressionXpathRegressionTrailingComment1 {
    int i; // do
}

$ java -jar checkstyle-8.36.2-all.jar -b "/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionTrailingComment1']]/OBJBLOCK/SINGLE_LINE_COMMENT/COMMENT_CONTENT" Test.java 
CLASS_DEF -> CLASS_DEF [3:0]
`--OBJBLOCK -> OBJBLOCK [3:56]
    |--SINGLE_LINE_COMMENT -> // [4:11]
    |   `--COMMENT_CONTENT ->  do\n [4:13]

$ java -jar checkstyle-8.36.2-all.jar -b "/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionTrailingComment1']]/OBJBLOCK/SINGLE_LINE_COMMENT/COMMENT_CONTENT[@text=' do\n']" Test.java 

$ java -jar checkstyle-8.36.2-all.jar -b "/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionTrailingComment1']]/OBJBLOCK/SINGLE_LINE_COMMENT/COMMENT_CONTENT[@text!=' do\n']" Test.java 
```

**Comment by romani** [comment]
```
$ cat Test.java 
public class Test {
    int i; // warn
    int[] arr = {
      40, // ok magic reason
      47, // ok magic reason
    };
}

$ java -jar checkstyle-8.36.2-all.jar -b "//ARRAY_INIT//SINGLE_LINE_COMMENT/COMMENT_CONTENT" Test.java 
CLASS_DEF -> CLASS_DEF [1:0]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--VARIABLE_DEF -> VARIABLE_DEF [3:7]
    |   |--ASSIGN -> = [3:14]
    |   |   `--ARRAY_INIT -> { [3:16]
    |   |       |--SINGLE_LINE_COMMENT -> // [5:10]
    |   |       |   `--COMMENT_CONTENT ->  ok magic reason\n [5:12]
---------
CLASS_DEF -> CLASS_DEF [1:0]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--VARIABLE_DEF -> VARIABLE_DEF [3:7]
    |   |--ASSIGN -> = [3:14]
    |   |   `--ARRAY_INIT -> { [3:16]
    |   |       |--EXPR -> EXPR [5:6]
    |   |       |   |--SINGLE_LINE_COMMENT -> // [4:10]
    |   |       |   |   `--COMMENT_CONTENT ->  ok magic reason\n [4:12]
```

**Comment by timurt** [comment]
@romani 

Generally, we do not support `@text` attribute for all token types, only for certain nodes that are described [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java#L101). 

I added `TokenTypes.COMMENT_CONTENT` token type to the list that I mentioned and assembled project to jar. After that I was able to query this `COMMENT_CONTENT` node by attribute value.

ex:
```
$ java -jar checkstyle-8.38-SNAPSHOT-all.jar -J Test.java -b "//COMMENT_CONTENT[starts-with(@text, ' do')]"
```

Here I used `starts-with` only because attribute contained escape character and I couldn't query it. I suppose we need to create a separate issue, if we want to handle `COMMENT_CONTENT` token type

**Comment by romani** [comment]
current implementation works like:
```
$ cat Test.java 
public class Test {
    int p; /* warn
              second line */
    int i; // warn
    int[] arr = {
      40, // ok magic reason
      47, // ok magic reason
    };
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="TrailingComment" />

      <module name="SuppressionXpathSingleFilter">
        <property name="checks" value="TrailingComment"/>
        <property name="query" value="//ARRAY_INIT//SINGLE_LINE_COMMENT"/>
      </module>

    </module>
</module>

$ java -jar checkstyle-8.38-SNAPSHOT-all.jar -c config.xml  Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:2:12: Don't use trailing comments. [TrailingComment]
[ERROR] /var/tmp/Test.java:4:12: Don't use trailing comments. [TrailingComment]
Audit done.
Checkstyle ends with 2 errors.

```

**Review by romani — APPROVED** [review]
I am ok to merge

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java [hunk 1]
```diff
@@ -201,14 +221,17 @@ public void beginTree(DetailAST rootAST) {
...
             if (cppComments.containsKey(lineNo)) {
                 comment = cppComments.get(lineNo);
                 lineBefore = line.substring(0, comment.getStartColNo());
+                commentColNo = comment.getStartColNo();
             }
             else {
                 final List<TextBlock> commentList = cComments.get(lineNo);
                 comment = commentList.get(commentList.size() - 1);
                 lineBefore = line.substring(0, comment.getStartColNo());
+                commentColNo = comment.getStartColNo();
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java:234]: `commentColNo = comment.getStartColNo();`
this does not depends on if condition, please inline its usage on line 245

**Review by strkkk — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java]: awesome feedback, fixed.

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
I just rebased (to fix wercker) after requesting report, code is the same.

**Review by strkkk — APPROVED** [review]

**Comment by romani** [comment]
Github, rebase

**Pull request merged**